src/java.base/share/classes/java/nio/Bits.java

```diff
@@ -118,29 +122,41 @@ static void reserveMemory(long size, long cap) {
             return;
         }
 
-        final JavaLangRefAccess jlra = SharedSecrets.getJavaLangRefAccess();
+        // Don't completely discard interruptions.  Instead, record them and
+        // reapply when we're done here (whether successfully or OOME).
         boolean interrupted = false;
         try {
-
-            // Retry allocation until success or there are no more
-            // references (including Cleaners that might free direct
-            // buffer memory) to process and allocation still fails.
-            boolean refprocActive;
-            do {
+            // Keep trying to reserve until either succeed or there is no
+            // further cleaning available from prior GCs. If the latter then
+            // GC to hopefully find more cleaning to do.
+            for (int cleanedEpoch = -1; true; ) {
+                synchronized (RESERVE_SLOWPATH_LOCK) {
+                    // Test if cleaning for prior GCs (from here) is complete.
+                    // If so, GC to produce more cleaning work, and increment
+                    // the counter to inform other threads that there may be
+                    // more cleaning work to do.  This is done under the lock
+                    // to close a race.  We could have multiple threads pass
+                    // the test "simultaneously", resulting in back-to-back
+                    // GCs.  For a STW GC the window is small, but for a
+                    // concurrent GC it's quite large.
+                    if (RESERVE_GC_EPOCH == cleanedEpoch) {
+                        // Increment with overflow to 0, so the value can
+                        // never equal the initial/reset cleanedEpoch value.
+                        RESERVE_GC_EPOCH = Integer.max(0, RESERVE_GC_EPOCH + 1);
```

Comment:
Could also do the following which avoids the branch in `Integer.max`:
```suggestion
                        RESERVE_GC_EPOCH = (RESERVE_GC_EPOCH + 1) & Integer.MAX_VALUE;
```

Reply:
I think the use of `Integer.max` is clearer, any performance difference is insignificant, and the compiler can
make that change as an optimization if appropriate. So I'm not inclined to accept this suggestion.